modules/date/date-parser.c

```diff
@@ -0,0 +1,226 @@
+static gboolean
+_parse_timestamp_and_deduce_missing_parts(DateParser *self, struct tm *tm, glong *tm_zone_offset, const gchar *input)
+{
+  gint current_year;
+  struct tm nowtm = *tm;
+  long tm_gmtoff;
+  const gchar *tm_zone = NULL;
+  const gchar *remainder;
+
+  current_year = tm->tm_year;
+  tm->tm_year = 0;
+  tm_gmtoff = -1;
+  remainder = strptime_with_tz(input, self->date_format, tm, &tm_gmtoff, &tm_zone);
+  if (!remainder || remainder[0])
+    return FALSE;
+
+  /* hopefully _parse_timestamp will fill the year information, if
+   * not, we are going to need the received year to find it out
+   * heuristically */
+
+  if (tm->tm_year == 0)
+    {
+      /* no year information in the timestamp, deduce it from the current year */
+      tm->tm_year = current_year;
+      tm->tm_year = determine_year_for_month(tm->tm_mon, &nowtm);
+    }
+
+  if (tm_gmtoff != -1)
+    *tm_zone_offset = tm_gmtoff;
+  else
+    *tm_zone_offset = -1;
+
+  return TRUE;
+}
+
+static void
+_adjust_tvsec_to_move_it_into_given_timezone(LogStamp *timestamp, gint normalized_hour, gint unnormalized_hour)
+{
+  timestamp->tv_sec = timestamp->tv_sec
+    + get_local_timezone_ofs(timestamp->tv_sec)
+    - (normalized_hour - unnormalized_hour) * 3600
+    - timestamp->zone_offset;
+}
+
+static glong
+_get_target_zone_offset(DateParser *self, glong tm_zone_offset, time_t now)
+{
+  if (tm_zone_offset != -1)
+    return tm_zone_offset;
+  else if (self->date_tz_info)
+    return time_zone_info_get_offset(self->date_tz_info, now);
+  else
+    return get_local_timezone_ofs(now);
+}
+
+static gboolean
+_convert_struct_tm_to_logstamp(DateParser *self, time_t now, struct tm *tm, glong tm_zone_offset, LogStamp *target)
+{
+  gint unnormalized_hour;
+
+  target->zone_offset = _get_target_zone_offset(self, tm_zone_offset, now);
+
+  /* NOTE: mktime changes struct tm in the call below! For instance it
+   * changes the hour value. (in daylight saving changes, and when it
+   * is out of range).
+   *
+   * We save the hour prior to this conversion, as it is needed when
+   * converting the timestamp from our local timezone to the specified
+   * one. */
+
+  /* FIRST: We convert the timestamp as it was in our local time zone. */
+  unnormalized_hour = tm->tm_hour;
+  target->tv_sec = cached_mktime(tm);
+
+  /* SECOND: adjust tv_sec as if we converted it according to our timezone. */
+  _adjust_tvsec_to_move_it_into_given_timezone(target, tm->tm_hour, unnormalized_hour);
+  return TRUE;
+}
+
+static gboolean
+_convert_timestamp_to_logstamp(DateParser *self, time_t now, LogStamp *target, const gchar *input)
+{
+  struct tm tm;
+  glong tm_zone_offset;
+
+  /* initialize tm with current date, this fills in dst and other
+   * fields (even non-standard ones) */
+
+  cached_localtime(&now, &tm);
+
+  if (!_parse_timestamp_and_deduce_missing_parts(self, &tm, &tm_zone_offset, input))
+    return FALSE;
+
+  if (!_convert_struct_tm_to_logstamp(self, now, &tm, tm_zone_offset, target))
```

Comment:
I could find both pro and con arguments for this. Let's see if there are other comments, if I do need to touch the code I'll address this az well.